rspack: concurrency issue when resolving the same virtual module

### Environment

`unplugin@v2.3.0`
`node@v22.13.1`

### Reproduction

I prepared a small repo to easily repro: https://github.com/elbywan/unplugin-repro

### Describe the bug

To handle virtual modules with rspack, unplugin creates a dummy file under the `node_modules/.virtual` folder.

But when a plugin resolves **multiple virtual imports to the same path in a very short notice** subsequent `resolveId` calls may return before the file is actually created - because they won't wait for the first `resolveId` to finish creating the vfs file - and it will cause the rspack resolver to complain that the file is missing.

### Additional context

_No response_

### Logs

This is the kind of error that is thrown:

```sh
ERROR in ./src/imports/7994.ts 1:0-26
  × Module not found: Can't resolve '/Users/julien.elbaz/dev/experiments/unplugin-repro/node_modules/.virtual/a%3Fvirtual' in '/Users/julien.elbaz/dev/experiments/unplugin-repro/src/imports'
   ╭────
 1 │ import a from 'a?virtual';
   ·               ───────────
   ╰────
```

--- a/src/rspack/index.ts
+++ b/src/rspack/index.ts
@@ -56,7 +56,7 @@ export function getRspackPlugin<UserOptions = Record<string, never>>(
           if (plugin.resolveId) {
             const vfs = new FakeVirtualModulesPlugin(plugin)
             vfs.apply(compiler)
-            plugin.__vfsModules = new Set()
+            plugin.__vfsModules = new Map()
             plugin.__vfs = vfs as any
 
             compiler.hooks.compilation.tap(plugin.name, (compilation, { normalModuleFactory }) => {
@@ -105,8 +105,14 @@ export function getRspackPlugin<UserOptions = Record<string, never>>(
                 // we treat it as a virtual module
                 if (!fs.existsSync(resolved)) {
                   if (!plugin.__vfsModules!.has(resolved)) {
-                    plugin.__vfsModules!.add(resolved)
-                    await vfs.writeModule(resolved)
+                    const fsPromise = vfs.writeModule(resolved)
+                    plugin.__vfsModules!.set(resolved, fsPromise)
+                    await fsPromise
+                  }
+                  else {
+                    // Ensure that the module is written to the virtual file system
+                    // before we use it.
+                    await plugin.__vfsModules!.get(resolved)
                   }
                   resolved = encodeVirtualModuleId(resolved, plugin)
                 }

--- a/src/types.ts
+++ b/src/types.ts
@@ -147,7 +147,7 @@ export interface UnpluginOptions {
 export interface ResolvedUnpluginOptions extends UnpluginOptions {
   // injected internal objects
   __vfs?: VirtualModulesPlugin
-  __vfsModules?: Set<string>
+  __vfsModules?: Map<string, Promise<string>>
   __virtualModulePrefix: string
 }
 

--- a/src/webpack/index.ts
+++ b/src/webpack/index.ts
@@ -54,7 +54,7 @@ export function getWebpackPlugin<UserOptions = Record<string, never>>(
               vfs = new VirtualModulesPlugin()
               compiler.options.plugins.push(vfs)
             }
-            plugin.__vfsModules = new Set()
+            plugin.__vfsModules = new Map()
             plugin.__vfs = vfs
 
             const resolverPlugin: ResolvePluginInstance = {
@@ -135,7 +135,7 @@ export function getWebpackPlugin<UserOptions = Record<string, never>>(
                       // https://github.com/unjs/unplugin/pull/155
                       if (!plugin.__vfsModules!.has(resolved)) {
                         plugin.__vfs!.writeModule(resolved, '')
-                        plugin.__vfsModules!.add(resolved)
+                        plugin.__vfsModules!.set(resolved, Promise.resolve(''))
                       }
                     }
 

